src/libpanic_unwind/emcc.rs

```diff
@@ -52,22 +52,32 @@ pub fn payload() -> *mut u8 {
     ptr::null_mut()
 }
 
+struct Exception {
+    data: Option<Box<dyn Any + Send>>,
+}
+
 pub unsafe fn cleanup(ptr: *mut u8) -> Box<dyn Any + Send> {
     assert!(!ptr.is_null());
-    let adjusted_ptr = __cxa_begin_catch(ptr as *mut libc::c_void);
-    let ex = ptr::read(adjusted_ptr as *mut _);
+    let adjusted_ptr = __cxa_begin_catch(ptr as *mut libc::c_void) as *mut Exception;
+    let ex = (*adjusted_ptr).data.take();
     __cxa_end_catch();
-    ex
+    ex.unwrap()
 }
 
 pub unsafe fn panic(data: Box<dyn Any + Send>) -> u32 {
     let sz = mem::size_of_val(&data);
-    let exception = __cxa_allocate_exception(sz);
-    if exception == ptr::null_mut() {
+    let exception = __cxa_allocate_exception(sz) as *mut Exception;
+    if exception.is_null() {
```

Comment:
AFAICT, this check is useless. I guess we might not want to change things in this PR though -- I base this on https://github.com/emscripten-core/emscripten/blob/incoming/system/lib/libcxxabi/src/cxa_exception.cpp#L180-L185 which never returns a null pointer here (instead aborting the process if we can't allocate). This is only notable insofar as the return type of `panic` on emscripten at least is always `!`, we never return a u32. I think we probably don't care though.

Reply:
I'm keeping this implementation conforming to the Itanium C++ ABI, so that it works on all platforms. I tested this code on Linux (i.e. using C++ exceptions as the backend for Rust panics).